src/time.c

```diff
@@ -39,13 +39,68 @@ rcutils_time_point_value_as_nanoseconds_string(
   if (0 == str_size) {
     return RCUTILS_RET_OK;
   }
-  if (rcutils_snprintf(str, str_size, "%.19" PRId64, *time_point) < 0) {
+  if (rcutils_snprintf(str, str_size, "%.19" PRId64, *time_point) <= 0) {
     RCUTILS_SET_ERROR_MSG("failed to format time point into string as nanoseconds");
     return RCUTILS_RET_ERROR;
   }
   return RCUTILS_RET_OK;
 }
 
+rcutils_ret_t
+rcutils_time_point_value_as_date_string(
+  const rcutils_time_point_value_t * time_point,
+  char * str,
+  size_t str_size)
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(time_point, RCUTILS_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(str, RCUTILS_RET_INVALID_ARGUMENT);
+  if (0 == str_size) {
+    return RCUTILS_RET_OK;
+  }
+  // best to abs it to avoid issues with negative values in C89, see:
+  //   https://stackoverflow.com/a/3604984/671658
+  uint64_t abs_time_point = (uint64_t)llabs(*time_point);
+  // break into two parts to avoid floating point error
+  uint64_t seconds = abs_time_point / (1000u * 1000u * 1000u);
+  uint64_t nanoseconds = abs_time_point % (1000u * 1000u * 1000u);
+  // Make sure the buffer is large enough to hold the largest possible uint64_t
+  char nanoseconds_str[21];
+
+  if (rcutils_snprintf(nanoseconds_str, sizeof(nanoseconds_str), "%" PRIu64, nanoseconds) <= 0) {
+    RCUTILS_SET_ERROR_MSG("failed to format time point nanoseconds into string");
+    return RCUTILS_RET_ERROR;
+  }
+
+  time_t now_t = (time_t)(seconds);
+  struct tm ptm = {.tm_year = 0, .tm_mday = 0};
+#ifdef _WIN32
+  if (localtime_s(&now_t, &ptm) == NULL) {
+    RCUTILS_SET_ERROR_MSG("failed to get localtime");
+    return RCUTILS_RET_ERROR;
+  }
+#else
+  if (localtime_r(&now_t, &ptm) == NULL) {
+    RCUTILS_SET_ERROR_MSG("failed to get localtime");
+    return RCUTILS_RET_ERROR;
+  }
+#endif
+
+  if (str_size < 32 || strftime(str, 32, "%Y-%m-%d %H:%M:%S", &ptm) == 0) {
+    RCUTILS_SET_ERROR_MSG("failed to format time point into string as iso8601_date");
+    return RCUTILS_RET_ERROR;
+  }
+  static const int date_end_position = 19;
+  if (rcutils_snprintf(
+      &str[date_end_position], str_size - date_end_position, ".%.3s",
+      nanoseconds_str) <= 0)
```

Comment:
```suggestion
      nanoseconds_str) < 0)
```

Reply:
@clalancette what do you think, I thought you agreed that checking <= 0 makes sense. When we write only zero bytes that seems like an error aswell

Reply:
Looking at the source for [rcutils_vsnprintf](https://github.com/ros2/rcutils/blob/0b8c10148a832f7ac5be0a9abdc10e7b73e4c51f/src/snprintf.c#L39-L72), it returns `-1` if it runs into an error (like the format being NULL).  Otherwise, at least on Linux, it returns the value that `vsnprintf` returns.  Looking at the man page for that, it says:

```
       The  functions  snprintf() and vsnprintf() do not write more than size bytes (including the terminating null byte ('\0')).  If the output was trun‐
       cated due to this limit, then the return value is the number of characters (excluding the terminating null byte) which would have been  written  to
       the  final  string  if enough space had been available.  Thus, a return value of size or more means that the output was truncated.  (See also below
       under NOTES.)

       If an output error is encountered, a negative value is returned.
```

So again on an error, a negative is returned.  So we definitely need to check for the negative value.

The `0` value as far as I can tell can't happen.  That's because of [this line](https://github.com/ros2/rcutils/blob/0b8c10148a832f7ac5be0a9abdc10e7b73e4c51f/src/snprintf.c#L55), where if the buffer_size was 0, we would have returned an error.

So I think that @fujitatomoya is right here and we only need the `<`.